design/FY2021/EquationFitCoilCurveReference.md

```diff
@@ -0,0 +1,111 @@
+﻿
+
+Change EquationFit coils to reference curve objects (Issue #8352)
+================
+
+**Yueyue Zhou, NREL**
+
+ - 11 Jan 2021
+  
+
+## Background ##
+
+Excerpts lifted from [issue #8352](https://github.com/NREL/EnergyPlus/issues/8352)
+
+ - The EquationFit coil objects (like `Coil:Cooling:WaterToAirHeatPump:EquationFit` and `Coil:Heating:WaterToAirHeatPump:EquationFit`) have curve coefficients embedded in them. 
+  
+ - But other coil objects refer to standalone curve objects instead.
+ 
+ - Embedded coefficients limit the flexibility of using EMS to model coil performance impact when needed.
+
+ - Changing EquationFit coils to reference curve objects also allows the user to use different equation forms and to use lookup tables. (From [Mike Witte](https://github.com/NREL/EnergyPlus/issues/8352#issuecomment-724140062))
+ 
+## E-mail and  Conference Call Conclusions ##
+
+*None yet.*
+
+## Coil Objects Need to change##
+
+- Coil:Cooling:WaterToAirHeatPump:EquationFit
+
+- Coil:Heating:WaterToAirHeatPump:EquationFit
+
+- HeatPump:WaterToWater:EquationFit:Cooling
+
+- HeatPump:WaterToWater:EquationFit:Heating
+
+## Curve Objects Needed ##
+- For `Coil:Cooling:WaterToAirHeatPump:EquationFit`, there're **quad-linear** curves and **quint-linear** curves used to describe coil performances. However, currently only **quad-linear** curves are existing.
+
+## Decision to Make ##
+
+### Add quint-linear curve object? ###
+As described above, both quad-linear and quint-linear curves are needed for  Coil:Cooling:WaterToAirHeatPump:EquationFit, while other coils only require quad-linear curves. In order to satisfy Coil:Cooling:WaterToAirHeatPump:EquationFit curve specifications, we can either:
+ - Add quint-linear curves so that the curve type of every curve referred can be explicitly specified. This is a more straightforward path to move forward to.
+ - Change quad-linear curves to be more general as "multi-linear" or simply replace it by quint-linear curves, regarding to the similarity between quad-linear and quint linear curves. Users can simply add 0 coefficients to whatever variables that are not important or E+ can only read specific number of coefficients for specific input. Notice that there's no tri-linear or bi-linear curve types in Energyplus which might imply some preference to this approach in its first place (Is that true? Any example of using quad-linear curve type to read in tri-linear curves?).
```

Comment:
My preference would be to have targeted curve types for each set of coefficients.  

Right now the CurveValue function takes 1 required independent variable and 5 `Objexx::Optional` arguments.  Then inside the function the values are passed into a curve worker function that then does operations to check if the arguments are present or not and override them with zeroes, then finally hit a big IF block on the curve type for evaluating the proper output.

The CurveValue function could eventually be refactored into individual functions that would reduce the amount of logic in each one and hopefully increase performance, especially considering how often curves are evaluated.  With that in mind, I think it would be better then to keep each curve type separate.

Reply:
Now here's an interesting aside.

The CurveValue function has 6 total possible independent variable [arguments](https://github.com/NREL/EnergyPlus/blob/6163e5b3936402b6de5ab05403cfd1696e85b5f9/src/EnergyPlus/CurveManager.cc#L163-L168).  1 is required and 5 are optional.  Inside that function, it switches depending on whether you are looking the value up in a table or evaluating a curve.  If you are using a table, it passes all 6 arguments.  However, if you are evaluating a curve, it only [passes](https://github.com/NREL/EnergyPlus/blob/6163e5b3936402b6de5ab05403cfd1696e85b5f9/src/EnergyPlus/CurveManager.cc#L203) the first 3!  Am I missing something?!

The condenser setpoint manager definitely uses a quad linear curve and [calls CurveValue with 4 independent variables](https://github.com/NREL/EnergyPlus/blob/6163e5b3936402b6de5ab05403cfd1696e85b5f9/src/EnergyPlus/SetPointManager.cc#L7064) expecting them to be used....is the 4th just ignored?   

Really though, am I missing something here @mjwitte @mitchute ?

Reply:
Yeah, this looks like a bug to me. However, even if it were passed, it looks like the only thing the fourth variable is used for in ```PerformanceCurveObject``` is to do bounds checking.

https://github.com/NREL/EnergyPlus/blob/6163e5b3936402b6de5ab05403cfd1696e85b5f9/src/EnergyPlus/CurveManager.cc#L2286

```V4``` doesn't appear in any of the ```CurveType``` equations.

Reply:
Nevermind, it does appear in the ```QuadLinear``` curve. So, yes, this is a bug.

Reply:
@Myoldmopar Um, no, it appears that's been broken since Curve:Quadlinear was first added - long, long, ago. A quick series of unit tests of curve types should verify that.

There are a couple of [rudimentary curve unit tests](https://github.com/NREL/EnergyPlus/blob/ec58f461aa80708e4ef10c7d96e489bed71d5965/tst/EnergyPlus/unit/CurveManager.unit.cc#L90) that don't even check the returned curve value, and this one has zero for the last two coefficients.

Reply:
Fantastic 🤦 

OK, so @yzhou601, for this feature to work, you'll need to modify the `CurveValue` logic to pass in more arguments.  There could be unintended consequences (performance) to this, so it's worth discussing how you want to handle this.  Do you want to open a separate issue and look at it yourself?  Open a separate issue and request someone else on the team address it?  Just address it as you work on this feature?

Reply:
Oops, a stray click marked this all as resolved for a bit. Sorry about that. The last time I was in there, I think I fixed one of the other less commonly used curves and it caused diffs. Maybe that won't happen this time, but it is a real possibility.

Reply:
@Myoldmopar Thanks, Edwin. As far as I know from the conversation, it doesn't sound like something complicated to address. I can try resolving it when I work on this feature, if I find any extra complication, I can try reaching out to one of you for some help. Does it sound good?

Reply:
I'm happy to take a look at this separately if needed.

Reply:
My suggestion is to hold off from doing a deep dive @mitchute, at least for now.  @yzhou601 go ahead and work on getting this patched and let's just hope for the best that it doesn't cause any unintended consequences.